quartz/build.ts

```diff
@@ -110,19 +126,148 @@ async function startServing(
     ignoreInitial: true,
   })
 
-  watcher
-    .on("add", (fp) => rebuildFromEntrypoint(fp, "add", clientRefresh, buildData))
-    .on("change", (fp) => rebuildFromEntrypoint(fp, "change", clientRefresh, buildData))
-    .on("unlink", (fp) => rebuildFromEntrypoint(fp, "delete", clientRefresh, buildData))
+  if (argv.fastRebuild) {
+    watcher
+      .on("add", (fp) => partialRebuild(fp, "add", clientRefresh, buildData))
+      .on("change", (fp) => partialRebuild(fp, "change", clientRefresh, buildData))
+      .on("unlink", (fp) => partialRebuild(fp, "delete", clientRefresh, buildData))
+  } else {
+    watcher
+      .on("add", (fp) => rebuildFromEntrypoint(fp, "add", clientRefresh, buildData))
+      .on("change", (fp) => rebuildFromEntrypoint(fp, "change", clientRefresh, buildData))
+      .on("unlink", (fp) => rebuildFromEntrypoint(fp, "delete", clientRefresh, buildData))
+  }
 
   return async () => {
     await watcher.close()
   }
 }
 
+async function partialRebuild(
+  filepath: string,
+  action: FileEvent,
+  clientRefresh: () => void,
+  buildData: BuildData, // note: this function mutates buildData
+) {
+  const { ctx, ignored, depGraphs, contentMap } = buildData
+  const { argv, cfg } = ctx
+  const toRemove = new Set<FilePath>()
+
+  // don't do anything for gitignored files
+  if (ignored(filepath)) {
+    return
+  }
+
+  const perf = new PerfTimer()
+  console.log(chalk.yellow("Detected change, rebuilding..."))
+
+  // UPDATE DEP GRAPH
+  const fp = joinSegments(argv.directory, toPosixPath(filepath)) as FilePath
+
+  const staticResources = getStaticResourcesFromPlugins(ctx)
+  let processedFile: ProcessedContent
+
+  switch (action) {
+    case "add":
+      // add to cache when new file is added
+      processedFile = (await parseMarkdown(ctx, [fp]))[0]
+      contentMap.set(fp, processedFile)
+
+      // update the dep graph by asking all emitters whether they depend on this file
+      for (const emitter of cfg.plugins.emitters) {
+        const emitterGraph = await emitter.getDependencyGraph(ctx, [processedFile], staticResources)
+        depGraphs[emitter.name].mergeEdgesForNode(emitterGraph, fp)
+      }
+      break
+    case "change":
+      // invalidate cache when file is changed
+      processedFile = (await parseMarkdown(ctx, [fp]))[0]
+      contentMap.set(fp, processedFile)
+
+      // only content files can have added/removed dependencies because of transclusions
+      if (path.extname(fp) === ".md") {
```

Comment:
technically media as well! people can embed images/videos/audio

Reply:
Ah, forgot about that, will include media embeds when I'm figuring out transclusions.

I think this check is correct though. We only need to check for `.md` because it is the only file type that can have dependencies.